pkg/build/sbom.go

```diff
@@ -26,42 +26,38 @@ import (
 	"istio.io/release-builder/pkg/util"
 )
 
-// Sbom generates Software Bill Of Materials for istio repo in an SPDX readable format.
+const (
+	sbomOutputURI string = "https://storage.googleapis.com/istio-release/releases"
+)
+
+// GenerateBillOfMaterials generates Software Bill Of Materials for istio repo in an SPDX readable format.
 func GenerateBillOfMaterials(manifest model.Manifest) error {
 	// Retrieve istio repository path to run the sbom generator
 	istioRepoDir := manifest.RepoDir("istio")
-	sourceSbomFile := path.Join(manifest.OutDir(), "istio-source.spdx")
-	sourceSbomNamespace := fmt.Sprintf("https://storage.googleapis.com/istio-release/releases/%s/istio-source.spdx",
-		manifest.Version)
-	releaseSbomFile := path.Join(manifest.OutDir(), "istio-release.spdx")
-	releaseSbomNamespace := fmt.Sprintf("https://storage.googleapis.com/istio-release/releases/%s/istio-release.spdx",
-		manifest.Version)
-
-	// construct all the docker image tarball names as bom currently cannot accept directory as input
-	dockerDir := path.Join(manifest.OutDir(), "docker")
-	dockerImages := []string{}
-	if err := filepath.Walk(dockerDir, func(path string, fi os.FileInfo, err error) error {
-		if err != nil {
-			return err
-		}
-		if fi == nil {
-			return fmt.Errorf("failed to get fileinfo for file at path %s", path)
-		}
-		if fi.IsDir() {
-			return nil
-		}
-		dockerImages = append(dockerImages, path)
-		return nil
-	}); err != nil {
-		return fmt.Errorf("failed to walk directory %s: %v", dockerDir, err)
+	nameSpaceURI := sbomOutputURI
+	if manifest.BillOfMaterialsURI != "" {
+		nameSpaceURI = manifest.BillOfMaterialsURI
 	}
+	sourceSbomFile := path.Join(manifest.OutDir(), fmt.Sprintf("istio-source-%s.spdx", manifest.Version))
+	sourceSbomNamespace := path.Join(nameSpaceURI, manifest.Version, fmt.Sprintf("istio-source-%s.spdx", manifest.Version))
+
+	releaseSbomFile := path.Join(manifest.OutDir(), fmt.Sprintf("istio-source-%s.spdx", manifest.Version))
+	releaseSbomNamespace := path.Join(nameSpaceURI, manifest.Version, fmt.Sprintf("istio-source-%s.spdx", manifest.Version))
 
 	// Run bom generator to generate the software bill of materials(SBOM) for istio.
 	log.Infof("Generating Software Bill of Materials for istio release artifacts")
-	if err := util.VerboseCommand("bom", "--log-level", "error", "generate", "--name", "Istio Release "+manifest.Version,
-		"--namespace", releaseSbomNamespace, "--ignore", "licenses,'*.sha256',docker", "--dirs", manifest.OutDir(),
-		"--image-archive", strings.Join(dockerImages, ","), "--output", releaseSbomFile).Run(); err != nil {
-		return fmt.Errorf("couldn't generate sbom for istio release artifacts: %v", err)
+	// For Docker output in 'context' mode we will not produce SBOM.
+	// `bom` can produce bill only for tar and remote images.
```

Comment:
@sergii-ssh  What happens when you build in "context" mode? Happy to learn about the use case if you think a feature is missing. Or is it for local development?

Reply:
hi @puerco In `context` mode images we build are stored in local registry/context, i.e. `gcr.io/istio-release/pilot:1.17.2`. Now if I run `bom generate -o istio-release.spdx --image gcr.io/istio-release/pilot:1.17.2` it will fail because image hasn't been pushed to remote registry yet.

```
FATA generating doc: creating SPDX document: generating SPDX package from image ref gcr.io/istio-release/pilot:1.17.2: while downloading images to archive: fetching remote descriptor: GET https://gcr.io/v2/istio-release/pilot/manifests/1.17.2: MANIFEST_UNKNOWN: Failed to fetch "1.17.2" from request "/v2/istio-release/pilot/manifests/1.17.2". 
```

Workaround is to `docker save` and then run with `--image-archive` but it would be nice to have something like `--image-local` flag to avoid doing so.

Reply:
Can we not point to the images that were created when in context (that's a bad name) mode? They won't been the final location. Or is the SBOM going to end up with inaccurate data? Maybe it would be accurate if we set `sbomOutputURI` to the correct value.

Reply:
SBOM location is configurable in `manifest.BillOfMaterialsURI` and filenames are stamped with version. Accuracy shouldn't be an issue unless spdx report overwriten on purpose.